Add dump command to allow container logs dumping

alexpusch opened the pull request:
A common use case in integration tests development and debugging is inspecting created container logs. 
Currently the way to do so is to run the tests using the "keep" command and inspect the logs afterwards. This isn't a good fit for CI testing environment, and might not be ideal for local development as well.

This PR adds a new "dump" command that removes the containers just like "remove" but will dump container logs into a "./testcontainers" directory.

The logs filename is `{container_name/image_name}_{stdout/stderr}_{current date in ISO format}.log`

usage example: `TESTCONTAINERS=dump cargo test`

thomaseizinger commented on testcontainers/src/core/container.rs:
```diff
@@ -227,12 +246,64 @@ where
+
+fn get_log_dump_dir_path() -> PathBuf {
+    env::current_dir()
+        .unwrap_or(PathBuf::from("/tmp"))
```
This is gonna fail on Windows. I'd say if we can't get the current directory, print a warning and lets not dump them.

alexpusch replied:
On second thought, it might be better to allow passing the required path from outside - maybe an additional env var, or some syntax on the command env var i.e `TESTCONTAINERS=dump=/tmp/target` ? what do you think?

thomaseizinger commented on testcontainers/src/core/logs.rs:
```diff
@@ -65,6 +65,10 @@ impl LogStream {
         Err(end_of_stream(lines))
     }
+
+    pub fn into_inner(self) -> Box<dyn Read> {
```
Is this now public API? Let's make it `pub(crate)` please.

thomaseizinger commented on testcontainers/src/core/container.rs:
```diff
@@ -227,12 +246,64 @@ where
     }
 }
 
+fn dump_logs<I>(container: &Container<'_, I>) -> io::Result<()>
```
This needs to be provided for `ContainerAsync` too.

thomaseizinger reviewed (commented):
Thanks! A few comments.

alexpusch commented:
@thomaseizinger glad you like the feature :) 
I added support for this in ContainerAsync as well.

In addition I changed to logs directory structure so that all logs from same test run will be saved in its own directory: `./testcontainers/{iso date}/{container name|image name}_{container_id}_{stdout|stderr}.log`. I think this would provide a better user experience.
This is done by saving the date of the first log dump in a lazy static variable.

DDtKey commented:
The ability to dump logs is definitely helpful

However I'd consider more aligned interface with implementations in other languages, like [java](https://java.testcontainers.org/features/container_logs/) and [golang](https://golang.testcontainers.org/features/follow_logs/)

Also keep in mind, that we gonna deprecate CLI client soon (#563). So we actually can introduce this for http client only

Subjectively, I'd avoid configuration though env variable and dumping logs to filesystem. Instead, it could be achieved by not-removing container and accessing logs with docker cli.

DDtKey commented:
Let us know if you're still interested in implementing / actualizing this PR.
We actively revamp the project
